src/transaction/mod.rs

```diff
@@ -53,52 +86,91 @@ pub enum Mutation {
     /// # let connected_client = connecting_client.await.unwrap();
     /// let mut txn = connected_client.begin().await.unwrap();
     /// let key = "TiKV".to_owned();
-    /// let req = txn.get(key);
-    /// let result: Value = req.await.unwrap();
+    /// let result: Option<Value> = txn.get(key).await.unwrap();
     /// // Finish the transaction...
     /// txn.commit().await.unwrap();
     /// # });
     /// ```
-    pub async fn get(&self, _key: impl Into<Key>) -> Result<Value> {
-        unimplemented!()
+    pub async fn get(&self, key: impl Into<Key>) -> Result<Option<Value>> {
+        let key = key.into();
+        match self.get_from_mutations(&key) {
+            MutationValue::Determined(value) => Ok(value),
+            MutationValue::Undetermined => self.snapshot.get(key).await,
+        }
     }
 
-    /// Gets the values associated with the given keys.
+    /// Gets the values associated with the given keys. The returned iterator is in the same order
+    /// as the given keys.
     ///
     /// ```rust,no_run
     /// # #![feature(async_await)]
-    /// # use tikv_client::{KvPair, Config, TransactionClient};
+    /// # use tikv_client::{Key, Value, Config, transaction::Client};
     /// # use futures::prelude::*;
+    /// # use std::collections::HashMap;
     /// # futures::executor::block_on(async {
-    /// # let connecting_client = TransactionClient::connect(Config::new(vec!["192.168.0.100", "192.168.0.101"]));
+    /// # let connecting_client = Client::connect(Config::new(vec!["192.168.0.100", "192.168.0.101"]));
     /// # let connected_client = connecting_client.await.unwrap();
     /// let mut txn = connected_client.begin().await.unwrap();
     /// let keys = vec!["TiKV".to_owned(), "TiDB".to_owned()];
-    /// let req = txn.batch_get(keys);
-    /// let result: Vec<KvPair> = req.await.unwrap();
+    /// let result: HashMap<Key, Value> = txn
+    ///     .batch_get(keys)
+    ///     .await
+    ///     .unwrap()
+    ///     .filter_map(|(k, v)| v.map(move |v| (k, v))).collect();
     /// // Finish the transaction...
     /// txn.commit().await.unwrap();
     /// # });
     /// ```
     pub async fn batch_get(
         &self,
-        _keys: impl IntoIterator<Item = impl Into<Key>>,
-    ) -> Result<Vec<KvPair>> {
-        unimplemented!()
+        keys: impl IntoIterator<Item = impl Into<Key>>,
+    ) -> Result<impl Iterator<Item = (Key, Option<Value>)>> {
+        let mut keys_from_snapshot = Vec::new();
+        let mut results_in_buffer = keys
+            .into_iter()
+            .map(|key| {
+                let key = key.into();
+                let mutation_value = self.get_from_mutations(&key);
+                if let MutationValue::Undetermined = mutation_value {
+                    keys_from_snapshot.push(key.clone());
+                }
+                (key, mutation_value)
+            })
+            .collect::<Vec<_>>()
+            .into_iter();
```

Comment:
There should be a better way than to `collect` then immediately iterate again. Maybe `cloned()` or add a `to_owned` to the map function? If you're doing this for the side effects, then please add a comment

Reply:
It's just an eager evaluation of the iterator. Then, the new iterator does not hold a reference to `self`.

Reply:
Yes, and there is side effect too. Now I think it's better to use a for loop...